pkg/kstatus/observe/observer/observer_test.go

```diff
@@ -0,0 +1,227 @@
+	"sigs.k8s.io/cli-utils/pkg/kstatus/status"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestStatusObserverRunner(t *testing.T) {
+	testCases := map[string]struct {
+		identifiers        []wait.ResourceIdentifier
+		defaultObserver    ResourceObserver
+		expectedEventTypes []event.EventType
+	}{
+		"no resources": {
+			identifiers:        []wait.ResourceIdentifier{},
+			expectedEventTypes: []event.EventType{event.CompletedEvent},
+		},
+		"single resource": {
+			identifiers: []wait.ResourceIdentifier{
+				{
+					GroupKind: schema.GroupKind{
+						Group: "apps",
+						Kind:  "Deployment",
+					},
+					Name:      "foo",
+					Namespace: "bar",
+				},
+			},
+			defaultObserver: &fakeObserver{
+				resourceObservations: map[schema.GroupKind][]status.Status{
+					schema.GroupKind{Group: "apps", Kind: "Deployment"}: { //nolint:gofmt
+						status.InProgressStatus,
+						status.CurrentStatus,
+					},
+				},
+				resourceObservationCount: make(map[schema.GroupKind]int),
+			},
+			expectedEventTypes: []event.EventType{
+				event.ResourceUpdateEvent,
+				event.ResourceUpdateEvent,
+				event.CompletedEvent,
+			},
+		},
+		"multiple resources": {
+			identifiers: []wait.ResourceIdentifier{
+				{
+					GroupKind: schema.GroupKind{
+						Group: "apps",
+						Kind:  "Deployment",
+					},
+					Name:      "foo",
+					Namespace: "default",
+				},
+				{
+					GroupKind: schema.GroupKind{
+						Group: "",
+						Kind:  "Service",
+					},
+					Name:      "bar",
+					Namespace: "default",
+				},
+			},
+			defaultObserver: &fakeObserver{
+				resourceObservations: map[schema.GroupKind][]status.Status{
+					schema.GroupKind{Group: "apps", Kind: "Deployment"}: { //nolint:gofmt
+						status.InProgressStatus,
+						status.CurrentStatus,
+					},
+					schema.GroupKind{Group: "", Kind: "Service"}: { //nolint:gofmt
+						status.InProgressStatus,
+						status.InProgressStatus,
+						status.CurrentStatus,
+					},
+				},
+				resourceObservationCount: make(map[schema.GroupKind]int),
+			},
+			expectedEventTypes: []event.EventType{
+				event.ResourceUpdateEvent,
+				event.ResourceUpdateEvent,
+				event.ResourceUpdateEvent,
+				event.ResourceUpdateEvent,
+				event.CompletedEvent,
+			},
+		},
+	}
+
+	for tn, tc := range testCases {
+		t.Run(tn, func(t *testing.T) {
+			ctx := context.Background()
+
+			identifiers := tc.identifiers
+
+			observer := Observer{
+				AggregatorFactoryFunc: func(identifiers []wait.ResourceIdentifier) StatusAggregator {
+					return newFakeAggregator(identifiers)
+				},
+				ReaderFactoryFunc: func(_ client.Reader, _ meta.RESTMapper, _ []wait.ResourceIdentifier) (ObserverReader, error) {
+					return testutil.NewNoopObserverReader(), nil
+				},
+				ObserversFactoryFunc: func(_ ObserverReader, _ meta.RESTMapper) (resourceObservers map[schema.GroupKind]ResourceObserver, defaultObserver ResourceObserver) {
```

Comment:
s                 o              l     o   n        g    t   hi s    l       i   n   e    is!